README.md
# Level9 Interpreter for Agon Light

Level9 is an interpreter for the games by the English text adventure company Level 9. These games received considerable
critical acclaim at the time and are probably the best text adventures written for the small cassette based 8-bit
computers common in Britain during the early to mid 1980s.

_Please note: This readme is specific to the port of the interpreter to Agon Light computer.<br>_
_**[README-original.md](README-original.md) is the original info from one of the interpreter authors, David Kinder.**_

---
**Table of contents**

- [About](#about)
- [About Agon Light port](#about-agon-light-port)
- [Usage](#usage)
- [Building](#building)
    - [Compilation prerequisites](#compilation-prerequisites)
    - [Making the binary](#making-the-binary)
- [Supported Games](#supported-games)
- [Tested games](#tested-games)
- [Links](#links)
- [License](#license)

**Additional local docs**

- [Original README of one of the authors](README-original.md)

---

# About

The Level9 interpreter was originally written by Glen Summers, and later extended by David Kinder, along with Alan
Staniforth, Simon Baldwin, Dieter Baron and Andreas Scherrer. It has the very useful ability to search for Level 9
game data in any file loaded into it, allowing it to find games in uncompressed memory dumps from emulators (such as
the commonly used Spectrum emulator SNA file format). Both the line-drawn graphics used in a number of Level 9's
earlier games and the bitmap graphics from later games are also supported.

# About Agon Light port

The port of the interpreter to Agon Light is rudimentary at the moment. While all
the code base is fully compiled and all features are supported, platform-specific
implementation uses project's generic code only. This also means no support
for graphics of any form.

The code uses standard MOS terminal, with all editing capabilities. However, backspace
seems to not work well at the moment.

# Usage

The interpreter is available as a simple binary. To run a given game, load the interpreter, switch
to the game directory and run the interpreter while indicating the game DAT file:

```
load level9.bin
cd level9/Jewels-of-Darkness_DOS_EN
run &040000 quest.dat
```

# Building

The game uses continuous memory model and as such would not fit in the 64k
of standard Z80 RAM address space without extra C code optimization. Luckily,
the Agon Light can utilize its eZ80's larger 24-bit address space and its 512k memory.
The game engine takes ~64kB of memory, the rest is available for the loaded game data, heap and stack.

## Compilation prerequisites

The game runs in eZ80 ADL mode and requires C compiler capable of producing such code.
The port was developed using AgDev LLVM compiler targeted for eZ80. Feel free to follow instructions
at [AgDev GitHub page](https://github.com/pcawte/AgDev) to install the suite. It is by itself
an adaptation of the [CE C/C++ Toolchain](https://ce-programming.github.io/toolchain/static/getting-started.html)
originally created for the TI-84 calculator.

_Please note: When placing AgDev on top of CEdev pack, don't try to build the latest version of C/CE++ from
the source, but use the provided downloadable stable distribution. The AgDev placed on top of the latest C/CE++
built from source works just fine, but creates incorrect Agon binaries at the moment (as of Oct 4th, 2023)._

## Making the binary

Agon-specific port `Makefile` is available in `Agon/` directory and if CEdev is installed,
simple `make clean && make` on Linux should produce the desired `dist/level9.bin` MOS binary.

```
cd Agon
make clean
make
```

_Please note: it is essential to run `make clean` as it links generic interpreter sources from the top directory
to the Agon directory for consumption by the LLVM compiler. Unfortunately, the current CE C/C++ Toolkit rules
are very strict and require all source files to be present at the top of compiled tree. The project structure
unfortunately prohibits us to compile from the project top (source files for other platforms would be picked up)._

# Supported Games

All games supported by the project should be working (V1 - V4), but the most of them have not been tested
in this Agon port. The best bet is to aim for DOS-based game distributions of Level 9 games coming in the form
of separate files. The interpreter requires a `*.dat` file to start, but will read other files when needed.
When grabbing files from other platforms, make sure you provide the actual data files to the interpreter,
not the disk or tape media for the given platform.

The interpreter also supports SNA snapshot files from ZX Spectrum as it will scan them for Level 9 structures.
This feature was not tested on Agon Light.

The interpreter does NOT support Spectrum Next ports of the games
available [here](https://github.com/stefanbylund/zxnext_level9/tree/master/compilation). These use specific
format of textures currently not supported by this main-derived code base of Level 9 interpreter.

# Tested games

**Jewel of Darkness DOS**

![Jewel of Darkness DOS on Agon Light](Agon/screenshots/jewel-of-darkness-dos-colossal.png)

# Links

## [Level 9 Interpreter](https://github.com/DavidKinder/Level9)

The original repository of the interpreter this port is based on.

## [Level 9 Memorial](http://l9memorial.if-legends.org/html/home.html)

Information about Level 9 Computing and their games.

## [Abandonware Level 9](https://www.myabandonware.com/browse/developer/level-9-computing-ltd-c2)

The original game disks are available in various forms on the internet, but might be a bit harder to find.
This is one of the places with better-organized files for download.

# License

The Level9 interpreter is licensed under the terms of the GNU General Public License version 2 and is copyright (C)
1996-2011 by Glen Summers and contributors. Contributions from David Kinder, Alan Staniforth, Simon Baldwin, Dieter
Baron and Andreas Scherrer.


CHANGELOG.md
# Changelog

All public releases of the Level 9 interpreter for Agon Light are documented in this file.

## 5.1.agon1 (2023-10-08)

**First public release**

- Original code base compiled for Agon Light, using generic, non-platform-specific routines
- No graphics is supported at the moment (neither vector mode 1, nor bitmap mode 2)
- Backspace seems to not be working correctly


level9.h
/***********************************************************************\
*
* Level 9 interpreter
* Version 5.2
* Copyright (c) 1996-2023 Glen Summers and contributors.
* Contributions from David Kinder, Alan Staniforth, Simon Baldwin,
* Dieter Baron and Andreas Scherrer.
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.
*
\***********************************************************************/

#ifdef HAVE_PLATFORM_H
#include "platform.h"
#endif

#ifndef L9TYPEDEF
#define L9TYPEDEF
typedef unsigned char L9BYTE;
typedef unsigned short L9UINT16;
typedef unsigned long L9UINT32;
typedef int L9BOOL;
#endif

#ifndef FALSE
#define FALSE 0
#define TRUE 1
#endif

#define LISTAREASIZE 0x800
#define STACKSIZE 1024
#define V1FILESIZE 0x600

#ifndef MAX_PATH
#define MAX_PATH 256
#endif

typedef struct
{
	L9UINT32 Id;
	L9UINT16 codeptr,stackptr,listsize,stacksize,filenamesize,checksum;
	L9UINT16 vartable[256];
	L9BYTE listarea[LISTAREASIZE];
	L9UINT16 stack[STACKSIZE];
	char filename[MAX_PATH];
} GameState;

typedef enum
{
	NO_BITMAPS,
	AMIGA_BITMAPS,
	PC1_BITMAPS,
	PC2_BITMAPS,
	C64_BITMAPS,
	BBC_BITMAPS,
	CPC_BITMAPS,
	MAC_BITMAPS,
	ST1_BITMAPS,
	ST2_BITMAPS,
} BitmapType;

typedef struct
{
	L9BYTE red, green, blue;
} Colour;

typedef struct
{
	L9UINT16 width, height;
	L9BYTE* bitmap;
	Colour palette[32];
	L9UINT16 npalette;
} Bitmap;

#define MAX_BITMAP_WIDTH 512
#define MAX_BITMAP_HEIGHT 218

#if defined(_Windows) || defined(__MSDOS__) || defined (_WIN32) || defined (__WIN32__)
	#define L9WORD(x) (*(L9UINT16*)(x))
	#define L9SETWORD(x,val) (*(L9UINT16*)(x)=(L9UINT16)val)
	#define L9SETDWORD(x,val) (*(L9UINT32*)(x)=val)
#else
	#define L9WORD(x) (*(x) + ((*(x+1))<<8))
	#define L9SETWORD(x,val) *(x)=(L9BYTE) val; *(x+1)=(L9BYTE)(val>>8);
	#define L9SETDWORD(x,val) *(x)=(L9BYTE)val; *(x+1)=(L9BYTE)(val>>8); *(x+2)=(L9BYTE)(val>>16); *(x+3)=(L9BYTE)(val>>24);
#endif

#if defined(_Windows) && !defined(__WIN32__)
#include <alloc.h>
#define malloc farmalloc
#define calloc farcalloc
#define free farfree
#endif

#ifdef __cplusplus
extern "C" {
#endif

/* routines provided by os dependent code */
void os_printchar(char c);
L9BOOL os_input(char* ibuff, int size);
char os_readchar(int millis);
L9BOOL os_stoplist(void);
void os_flush(void);
L9BOOL os_save_file(L9BYTE* Ptr, int Bytes);
L9BOOL os_load_file(L9BYTE* Ptr, int* Bytes, int Max);
L9BOOL os_get_game_file(char* NewName, int Size);
void os_set_filenumber(char* NewName, int Size, int n);
void os_graphics(int mode);
void os_cleargraphics(void);
void os_setcolour(int colour, int index);
void os_drawline(int x1, int y1, int x2, int y2, int colour1, int colour2);
void os_fill(int x, int y, int colour1, int colour2);
void os_show_bitmap(int pic, int x, int y);
FILE* os_open_script_file(void);
L9BOOL os_find_file(char* NewName);

/* routines provided by level9 interpreter */
L9BOOL LoadGame(char* filename, char* picname);
L9BOOL RunGame(void);
void StopGame(void);
void RestoreGame(char* filename);
void FreeMemory(void);
void GetPictureSize(int* width, int* height);
L9BOOL RunGraphics(void);

/* bitmap routines provided by level9 interpreter */
BitmapType DetectBitmaps(char* dir);
Bitmap* DecodeBitmap(char* dir, BitmapType type, int num, int x, int y);

#ifdef __cplusplus
}
#endif


Agon/platform.h
/***********************************************************************
*
* Level 9 interpreter
* Version 5.2.agon1
* Copyright (c) 1996-2023 Glen Summers and contributors.
* Contributions from David Kinder, Alan Staniforth, Simon Baldwin,
* Dieter Baron and Andreas Scherrer.
*
* Agon 8-bit ez80 version (w/ 24-bit addressing) by Roman Dolejsi.
* Platform-specific header file.
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.
*
\***********************************************************************/

#include <stdbool.h>
#include <mos_api.h>

#ifndef L9TYPEDEF
#define L9TYPEDEF
typedef uint8_t L9BYTE;
typedef uint16_t L9UINT16;
typedef uint32_t L9UINT32;
typedef bool L9BOOL;
#endif


Agon/agon.c
/***********************************************************************\
*
* Level 9 interpreter
* Version 5.2
* Copyright (c) 1996-2023 Glen Summers and contributors.
* Contributions from David Kinder, Alan Staniforth, Simon Baldwin,
* Dieter Baron and Andreas Scherrer.
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.
*
\***********************************************************************/

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#include "level9.h"

#define GFX_FILE_EXT_DOS_HIRES ".hrc"
#define GFX_FILE_EXT_DOS_CGA ".cga"
#define GFX_FILE_DEFAULT "picture.dat"

// Graphics is not enabled at the moment as it is not supported yet.
// Not only the graphical methods will need to be implemented, but also getchar needs to be redesigned
// to not wait for next character, but just check the key pressed (to allow graphics ticking).
#undef GFX_ENABLED

// When graphics is enabled, realtime keystrokes are grabbed and typical keyboard behavior is emulated.
// This includes initial and subsequent delays on key press to avoid too fast typing.
#define KEY_PRESS_DELAY_TICKS_INITIAL 7500
#define KEY_PRESS_DELAY_TICKS_SUBSEQUENT 3500

#define TEXTBUFFER_SIZE 10240
char text_buffer[TEXTBUFFER_SIZE + 1];
int text_buffer_pointer = 0;

uint8_t column = 0;
static uint16_t screen_width;
static uint16_t screen_height;
static uint8_t screen_columns;
static uint8_t screen_rows;
static uint8_t screen_colors;

/**
 * Loads the current graphic mode details for future use.
 */
void os_graphics_inquiry() {
    screen_width = getsysvar_scrwidth();
    screen_height = getsysvar_scrheight();
    screen_columns = getsysvar_scrCols();
    screen_rows = getsysvar_scrRows();
    screen_colors = getsysvar_scrColours();
}

void os_printchar(char c) {
    if (c == '\r') {
        os_flush();
        putchar('\n');
        column = 0;
    } else if (isprint(c) != 0) {
        if (text_buffer_pointer >= TEXTBUFFER_SIZE) {
            os_flush();
        }
        *(text_buffer + (text_buffer_pointer++)) = c;
    }
}

char os_key_pressed() {
    static uint8_t prev_down = 0;
    static uint16_t down_delay = 0;

    // check if anything is pressed, clear previous pressed counter
    uint8_t down = getsysvar_vkeydown();
    if (!down) {
        prev_down = down;
        return 0;
    }

    // get realtime key and correct it for stdio compatibility
    uint8_t key = getsysvar_keyascii();
    if (key == '\r') {
        key = '\n';
    }

    // key just pressed, set delay to first delay, return the keystroke
    if (!prev_down && down) {
        prev_down = down;
        down_delay = KEY_PRESS_DELAY_TICKS_INITIAL;
        return key;
    }

    // we are waiting until next key repeat
    if (--down_delay > 0) {
        return 0;
    }

    // set follow-up key delay, return the key
    down_delay = KEY_PRESS_DELAY_TICKS_SUBSEQUENT;
    return key;
}

char os_key_next() {
    static uint8_t prev_down = 0;
    static uint16_t down_delay = 0;

    // check if anything is pressed, clear previous pressed counter
    uint8_t down = getsysvar_vkeydown();
    if (!down) {
        prev_down = down;
        return 0;
    }

    // get realtime key and correct it for stdio compatibility
    uint8_t key = getsysvar_keyascii();
    if (key == '\r') {
        key = '\n';
    }

    // key just pressed, set delay to first delay, return the keystroke
    if (!prev_down && down) {
        prev_down = down;
        down_delay = KEY_PRESS_DELAY_TICKS_INITIAL;
        return key;
    }

    // we are waiting until next key repeat
    if (--down_delay > 0) {
        return 0;
    }

    // set follow-up key delay, return the key
    down_delay = KEY_PRESS_DELAY_TICKS_SUBSEQUENT;
    return key;
}

/**
 * Corrects delete character (shifts remaining content left by one, removing previous character + delete).
 */
void os_correct_input_deletes(char *ibuff, int size) {
    int src_pos = 0;
    int dst_pos = 0;
    char ch;
    do {
        ch = ibuff[src_pos++];
        if (ch != 127) {
            ibuff[dst_pos++] = ch;
        } else {
            if (dst_pos > 0) {
                dst_pos--;
            }
        }
    } while (ch != 0 && src_pos < size);
    // the terminating character should've been written, but let's finalize the string anyway
    ibuff[dst_pos] = 0;
}

L9BOOL os_input(char *ibuff, int size) {
    os_flush();
#ifdef GFX_ENABLED
    char ch;
    int ibuff_pos = 0;
    do {
        ch = os_key_pressed();
        if (ch != 0) {
            if (ch == 127) { // delete
                if (ibuff_pos > 0) {
                    ibuff[ibuff_pos--] = 0;
                    // print delete (remove character from console)
                    printf("%c", ch);
                }
            } else {
                ibuff[ibuff_pos++] = ch;
                printf("%c", ch);
            }
        }
        RunGraphics();
    } while (ch != '\n' && ibuff_pos < size);
#else
    fgets(ibuff, size, stdin);
    os_correct_input_deletes(ibuff, size);
#endif
    char *nl = strchr(ibuff, '\n');
    if (nl) {
        *nl = 0;
    }
    return TRUE;
}

char os_readchar(int millis) {
    os_flush();
#ifdef GFX_ENABLED
    RunGraphics();
#endif
    if (millis == 0) {
        return 0;
    }

    /* Some of the Level 9 games expect to be able to wait for
       a character for a short while as a way of pausing, and
       expect 0 to be returned, while the multiple-choice games
       (such as The Archers) expect 'proper' keys from this
       routine.

       To get round this, we return 0 for the first 1024 calls,
       and 'proper' keys thereafter. Since The Archers and
       similar games ignore the returned zeros, this works quite
       well. A 'correct' port would solve this properly by
       implementing a timed wait for a key, but this is not
       possible using only C stdio-functions.
    */
    static int count = 0;
    if (++count < 1024) {
        return 0;
    }
    count = 0;

    char c;
#ifdef GFX_ENABLED
    /* in realtime mode we just wait for next key, nothing to clear from any stdin buffer */
    do {
        RunGraphics();
        c = os_key_pressed();
    } while (c == 0);
#else
    c = getc(stdin); /* will require enter key as well */
    if (c != '\n') {
        while (getc(stdin) != '\n') {
            /* remove input from buffer until enter key */
        }
    }
#endif
    return c;
}

L9BOOL os_stoplist(void) {
    return FALSE;
}

void os_flush(void) {
    if (text_buffer_pointer < 1) {
        return;
    }
    *(text_buffer + text_buffer_pointer) = ' ';
    int ptr = 0;
    while (text_buffer_pointer + column > screen_columns - 1) {
        register int space = ptr;
        register int last_space = space;
        register bool searching = TRUE;
        while (searching) {
            while (text_buffer[space] != ' ') {
                space++;
            }
            if (space - ptr + column > screen_columns - 1) {
                space = last_space;
                printf("%.*s\n", space - ptr, text_buffer + ptr);
                column = 0;
                space++;
                if (text_buffer[space] == ' ') {
                    space++;
                }
                text_buffer_pointer -= (space - ptr);
                ptr = space;
                searching = FALSE;
            } else {
                last_space = space;
            }
            space++;
        }
    }
    if (text_buffer_pointer > 0) {
        printf("%.*s", text_buffer_pointer, text_buffer + ptr);
        column += text_buffer_pointer;
    }
    text_buffer_pointer = 0;
}

FILE *os_prompt_and_open_file(const char *prompt, const char *mode) {
    char name[256];
    os_flush();
    printf("%s", prompt);
    fgets(name, 256, stdin);
    char *nl = strchr(name, '\n');
    if (nl) {
        *nl = 0;
    }
    return fopen(name, mode);
}

L9BOOL os_save_file(L9BYTE *Ptr, int Bytes) {
    FILE *f = os_prompt_and_open_file("Save file: ", "wb");
    if (!f) {
        return FALSE;
    }
    fwrite(Ptr, 1, Bytes, f);
    fclose(f);
    return TRUE;
}

L9BOOL os_load_file(L9BYTE *pointer, int *bytes, int max) {
    FILE *f = os_prompt_and_open_file("Load file: ", "rb");
    if (!f) {
        return FALSE;
    }
    *bytes = fread(pointer, 1, max, f);
    fclose(f);
    return TRUE;
}

L9BOOL os_get_game_file(char *new_name, int size) {
    os_flush();
    printf("Load next game: ");
    fgets(new_name, size, stdin);
    char *nl = strchr(new_name, '\n');
    if (nl) {
        *nl = 0;
    }
    return TRUE;
}

void os_set_filenumber(char *new_name, int size, int n) {
    char *p = strrchr(new_name, '/');
    if (p == NULL) {
        p = new_name;
    }
    for (int i = strlen(p) - 1; i >= 0; i--) {
        if (isdigit(p[i])) {
            p[i] = '0' + n;
            return;
        }
    }
}

FILE *os_open_script_file(void) {
    return os_prompt_and_open_file("Script file: ", "rt");
}

L9BOOL os_find_file(char *file_name) {
    FILE *f = fopen(file_name, "rb");
    if (f != NULL) {
        fclose(f);
        return TRUE;
    }
    return FALSE;
}

void os_graphics(int mode) {
#ifdef GFX_DEBUG
    printf("os_graphics(mode=%i);\n", mode);
#endif
}

void os_cleargraphics(void) {
#ifdef GFX_DEBUG
    printf("os_cleargraphics();\n");
#endif
}

void os_setcolour(int color, int index) {
#ifdef GFX_DEBUG
    printf("os_setcolour(color=%i,index=%i);\n", color, index);
#endif
}

void os_drawline(int x1, int y1, int x2, int y2, int color1, int color2) {
#ifdef GFX_DEBUG
    printf("os_drawline(x1=%i,y1=%i,x2=%i,y2=%i,color1=%i,color2=%i);\n", x1, y1, x2, y2, color1, color2);
#endif
}

void os_fill(int x, int y, int color1, int color2) {
#ifdef GFX_DEBUG
    printf("os_fill(x=%i,y=%i,color1=%i,color2=%i);\n", x, y, color1, color2);
#endif
}

void os_show_bitmap(int pic, int x, int y) {
#ifdef GFX_DEBUG
    printf("os_show_bitmap(pic=%i,x=%i,y=%i);\n", pic, x, y);
#endif
}

char *os_strip_file_ext(char *file) {
    char *ext_pos = strrchr(file, '.');
    return ext_pos != NULL ? strndup(file, strlen(file) - strlen(ext_pos)) : file;
}

/**
 * Performs a secondary graphics file lookup.
 *
 * We currently support DOS-based games coming with *.cga or *.hrc graphics,
 * where Hi-Res graphics takes precedence unless number of screen colors or resolution is below 512.
 * The graphics file can be forced from command line, though - irrespective of the screen limitations.
 *
 * @return either gfx file forced from command line or detected file based on screen preference
 */
void os_lookup_gfx_file(char *dat_file, char *gfx_file_forced, char *gfx_file) {
    if (gfx_file_forced != NULL) {
        strcpy(gfx_file, gfx_file_forced);
        return;
    }
    printf("No gfx file specified, detecting..\n");
    char *file_basename = os_strip_file_ext(dat_file);

    char dos_hires[256];
    sprintf(dos_hires, "%s%s", file_basename, GFX_FILE_EXT_DOS_HIRES);
    L9BOOL dos_hires_exists = os_find_file(dos_hires);

    char dos_cga[256];
    sprintf(dos_cga, "%s%s", file_basename, GFX_FILE_EXT_DOS_CGA);
    L9BOOL dos_cga_exists = os_find_file(dos_cga);

    L9BOOL default_exists = os_find_file(GFX_FILE_DEFAULT);

    // no gfx file detected
    gfx_file[0] = '\0'; // empty file
    if (!dos_cga_exists && !dos_hires_exists && !default_exists) {
        printf("No gfx file exists.\n");
        return;
    }
    // default file detected
    if (!dos_cga_exists && !dos_hires_exists) {
        printf("Using default gfx file %s (no other variants available).\n", GFX_FILE_DEFAULT);
        strcpy(gfx_file, GFX_FILE_DEFAULT);
        return;
    }
    // hires for good screen conditions or if cga does not exist
    if (dos_hires_exists && ((screen_width >= 512 && screen_colors >= 16) || !dos_cga_exists)) {
        printf("Using dos-hi-res gfx file %s.\n", GFX_FILE_DEFAULT);
        strcpy(gfx_file, dos_hires);
        return;
    }
    // CGA variant for everything else (320 width screen, 4 colors, ..)
    printf("Using dos-cga gfx file %s.\n", dos_cga);
    strcpy(gfx_file, dos_cga);
}

int main(int argc, char **argv) {
    os_graphics_inquiry();
    printf(
            "Level 9 Interpreter, mode %ix%i (%ix%i chars), %i colors\n",
            screen_width, screen_height, screen_columns, screen_rows, screen_colors
    );
    if (argc != 2 && argc != 3) {
        printf("Use: %s <game_file> [gfx_file]\n", argv[0]);
        return 0;
    }
    char gfx_file[256] = "";
    os_lookup_gfx_file(argv[1], argc < 3 ? NULL : argv[2], gfx_file);
    if (strlen(gfx_file) == 0) {
        printf("Loading game file %s\n", argv[1]);
    } else {
        printf("Loading game file %s + gfx file %s\n", argv[1], gfx_file);
    }
    if (!LoadGame(argv[1], gfx_file)) {
        printf("Error: Unable to open game\n");
        return 0;
    }
    while (RunGame());
    StopGame();
    FreeMemory();
    return 0;
}
